pkg/cloudprovider/vsphereparavirtual/route.go

```diff
@@ -0,0 +1,415 @@
+	klog.V(6).Infof("Attempting to list Routes for cluster %s", clusterName)
+
+	routeSets, err := r.routeClient.NsxV1alpha1().RouteSets(r.namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return []*cloudprovider.Route{}, nil
+		}
+		klog.ErrorS(ErrListRouteSet, fmt.Sprintf("%v", err))
+		return nil, err
+	}
+	if len(routeSets.Items) == 0 {
+		return []*cloudprovider.Route{}, nil
+	}
+	return r.createCPRoutes(routeSets, clusterName), nil
+}
+
+// createCPRoutes creates cloudprovider Routes based on RouteSet CR
+func (r *routesProvider) createCPRoutes(routeSets *v1alpha1.RouteSetList, clusterName string) []*cloudprovider.Route {
+	var routes []*cloudprovider.Route
+	for _, routeSet := range routeSets.Items {
+		routeSetClusterName := routeSet.Annotations[AnnotationClusterName]
+		if routeSetClusterName == clusterName {
+			// only return cloudprovider.Route if RouteSet CR status 'Ready' is true
+			// condition := GetRouteSetCondition(&(routeSet.Status), v1alpha1.RouteSetConditionTypeReady)
+			// if condition != nil && condition.Status == v1.ConditionTrue {
+			// one RouteSet per node, so we can use nodeName as the name of RouteSet CR
+			nodeName := routeSet.Name
+			for _, route := range routeSet.Spec.Routes {
+				cpRoute := &cloudprovider.Route{
+					Name:            route.Name,
+					TargetNode:      types.NodeName(nodeName),
+					DestinationCIDR: route.Destination,
+				}
+				routes = append(routes, cpRoute)
+			}
+			// }
+		}
+	}
+	return routes
+}
+
+// CreateRoute implements Routes.CreateRoute
+// Create a RouteSet custom resource for a Node
+func (r *routesProvider) CreateRoute(ctx context.Context, clusterName string, nameHint string, route *cloudprovider.Route) error {
+	nodeName := string(route.TargetNode)
+	klog.V(6).Infof("Creating Route for node %s with hint %s in cluster %s", nodeName, nameHint, clusterName)
+
+	nodeIP, err := r.getNodeIPAddress(nodeName, util.IsIPv4(route.DestinationCIDR))
+	if err != nil {
+		klog.Errorf("getting node %s IP address failed: %v", nodeName, err)
+		return err
+	}
+
+	_, err = r.createRouteSetCR(ctx, clusterName, nameHint, nodeName, route.DestinationCIDR, nodeIP)
+	if err != nil {
+		klog.Errorf("creating RouteSet CR for node %s failed: %s", nodeName, err)
+		return err
+	}
+	return nil
+	// TODO: Get realized state of corresponding static route
+	// return r.checkStaticRouteRealizedState(routeSet.Name)
+}
+
+// createRouteSetCR creates RouteSet CR through RouteSet client
+func (r *routesProvider) createRouteSetCR(ctx context.Context, clusterName string, nameHint string, nodeName string, cidr string, nodeIP string) (*v1alpha1.RouteSet, error) {
+	annotations := map[string]string{
+		AnnotationClusterName: clusterName,
+	}
+	nodeRef := metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        types.UID(nameHint),
+	}
+	r.ownerRefs = append(r.ownerRefs, nodeRef)
```

Comment:
Should it be a tmp ownerRef other than directly assigning to this slice used for all nodes?

Reply:
ah good catch, let me fix it